File: packages/faro-ios-sdk/Sources/FaroSDK/Persistence/DiskBuffer.swift
import Foundation

public struct DiskBufferConfig {
    public let maxDiskUsageBytes: Int64
    public let maxFileAgeSeconds: TimeInterval
    public let signalsDir: String
    public let crashDir: String

    public init(
        maxDiskUsageBytes: Int64 = 5 * 1024 * 1024,
        maxFileAgeSeconds: TimeInterval = 24 * 60 * 60,
        signalsDir: String = "faro_signals",
        crashDir: String = "faro_crashes"
    ) {
        self.maxDiskUsageBytes = maxDiskUsageBytes
        self.maxFileAgeSeconds = maxFileAgeSeconds
        self.signalsDir = signalsDir
        self.crashDir = crashDir
    }
}

internal final class DiskBuffer {
    private let signalsDir: URL
    private let crashDir: URL
    private let config: DiskBufferConfig
    private let logger: InternalLogger
    private let fileManager = FileManager.default
    private let writeQueue = DispatchQueue(label: "com.grafana.faro.diskbuffer")

    init(baseDir: URL, config: DiskBufferConfig, logger: InternalLogger) {
        self.config = config
        self.logger = logger
        self.signalsDir = baseDir.appendingPathComponent(config.signalsDir)
        self.crashDir = baseDir.appendingPathComponent(config.crashDir)

        try? fileManager.createDirectory(at: signalsDir, withIntermediateDirectories: true)
        try? fileManager.createDirectory(at: crashDir, withIntermediateDirectories: true)
    }

    func writeSignal(_ body: TransportBody) {
        writeQueue.async { [weak self] in
            guard let self = self else { return }
            self.enforceStorageLimits()

            let fileName = "\(Int(Date().timeIntervalSince1970 * 1000))_\(UUID().uuidString).json"
            let fileURL = self.signalsDir.appendingPathComponent(fileName)

            do {
                let data = try body.toJSON()
                try data.write(to: fileURL)
                self.logger.debug("Signal written to disk: \(fileName)")
            } catch {
                self.logger.error("Failed to write signal to disk", error: error)
            }
        }
    }

    func writeCrash(_ body: TransportBody) {
        // Write synchronously for crash scenarios
        let fileName = "crash_\(Int(Date().timeIntervalSince1970 * 1000))_\(UUID().uuidString).json"
        let fileURL = crashDir.appendingPathComponent(fileName)

        do {
            let data = try body.toJSON()
            try data.write(to: fileURL, options: .atomic)
        } catch {
            // Best effort in crash handler
        }
    }

    func readPendingSignals() -> [SignalFile] {
        return readFiles(from: signalsDir)
    }

    func readPendingCrashes() -> [SignalFile] {
        return readFiles(from: crashDir)
    }

    private func readFiles(from directory: URL) -> [SignalFile] {
        guard let fileURLs = try? fileManager.contentsOfDirectory(
            at: directory,
            includingPropertiesForKeys: [.nameKey],
            options: .skipsHiddenFiles
        ) else {
            return []
        }

        return fileURLs
            .filter { $0.pathExtension == "json" }
            .sorted { $0.lastPathComponent < $1.lastPathComponent }
            .compactMap { url in
                do {
                    let data = try Data(contentsOf: url)
                    let body = try TransportBody.fromJSON(data)
                    return SignalFile(url: url, body: body)
                } catch {
                    logger.error("Failed to read signal file: \(url.lastPathComponent)", error: error)
                    try? fileManager.removeItem(at: url)
                    return nil
                }
            }
    }

    func deleteFile(_ signalFile: SignalFile) {
        try? fileManager.removeItem(at: signalFile.url)
    }

    func enforceStorageLimits() {
        let now = Date()

        // Delete old files
        let allFiles = getAllFileURLs()
        for fileURL in allFiles {
            if let attrs = try? fileManager.attributesOfItem(atPath: fileURL.path),
               let modDate = attrs[.modificationDate] as? Date,
               now.timeIntervalSince(modDate) > config.maxFileAgeSeconds {
                try? fileManager.removeItem(at: fileURL)
                logger.debug("Deleted expired signal file: \(fileURL.lastPathComponent)")
            }
        }

        // Enforce total size limit
        var totalSize: Int64 = 0
        let sortedFiles = getAllFileURLs().sorted { url1, url2 in
            let date1 = (try? fileManager.attributesOfItem(atPath: url1.path)[.modificationDate] as? Date) ?? .distantPast
            let date2 = (try? fileManager.attributesOfItem(atPath: url2.path)[.modificationDate] as? Date) ?? .distantPast
            return date1 < date2
        }

        for fileURL in sortedFiles {
            if let attrs = try? fileManager.attributesOfItem(atPath: fileURL.path),
               let size = attrs[.size] as? Int64 {
                totalSize += size
            }
        }

        if totalSize > config.maxDiskUsageBytes {
            for fileURL in sortedFiles {
                guard totalSize > config.maxDiskUsageBytes else { break }
                if let attrs = try? fileManager.attributesOfItem(atPath: fileURL.path),
                   let size = attrs[.size] as? Int64 {
                    try? fileManager.removeItem(at: fileURL)
                    totalSize -= size
                    logger.debug("Deleted signal file to enforce size limit: \(fileURL.lastPathComponent)")
                }
            }
        }
    }

    private func getAllFileURLs() -> [URL] {
        let signalFiles = (try? fileManager.contentsOfDirectory(
            at: signalsDir, includingPropertiesForKeys: nil, options: .skipsHiddenFiles
        )) ?? []
        let crashFiles = (try? fileManager.contentsOfDirectory(
            at: crashDir, includingPropertiesForKeys: nil, options: .skipsHiddenFiles
        )) ?? []
        return signalFiles + crashFiles
    }

    func clear() {
        let allFiles = getAllFileURLs()
        for fileURL in allFiles {
            try? fileManager.removeItem(at: fileURL)
        }
    }
}

File: packages/faro-android-sdk/src/main/kotlin/com/edkimmel/faro/persistence/DiskBuffer.kt
package com.edkimmel.faro.persistence

import com.edkimmel.faro.internal.InternalLogger
import com.edkimmel.faro.transport.TransportBody
import java.io.File
import java.util.UUID

data class DiskBufferConfig(
    val maxDiskUsageBytes: Long = 5L * 1024 * 1024, // 5MB
    val maxFileAgeMs: Long = 24L * 60 * 60 * 1000,  // 24 hours
    val crashDir: String = "faro_crashes",
    val signalsDir: String = "faro_signals"
)

internal class DiskBuffer(
    private val baseDir: File,
    private val config: DiskBufferConfig,
    private val logger: InternalLogger
) {
    private val signalsDir = File(baseDir, config.signalsDir).apply { mkdirs() }
    private val crashDir = File(baseDir, config.crashDir).apply { mkdirs() }

    fun writeSignal(body: TransportBody) {
        try {
            enforceStorageLimits()
            val fileName = "${System.currentTimeMillis()}_${UUID.randomUUID()}.json"
            val file = File(signalsDir, fileName)
            file.writeText(TransportBody.toJson(body))
            logger.debug("Signal written to disk: ${file.name}")
        } catch (e: Exception) {
            logger.error("Failed to write signal to disk", e)
        }
    }

    fun writeCrash(body: TransportBody) {
        try {
            val fileName = "crash_${System.currentTimeMillis()}_${UUID.randomUUID()}.json"
            val file = File(crashDir, fileName)
            // Write synchronously for crash scenarios
            file.writeText(TransportBody.toJson(body))
        } catch (e: Exception) {
            // Best effort - we're in a crash handler
        }
    }

    fun readPendingSignals(): List<SignalFile> {
        return readFilesFromDir(signalsDir)
    }

    fun readPendingCrashes(): List<SignalFile> {
        return readFilesFromDir(crashDir)
    }

    private fun readFilesFromDir(dir: File): List<SignalFile> {
        return dir.listFiles()
            ?.filter { it.isFile && it.extension == "json" }
            ?.sortedBy { it.name }
            ?.mapNotNull { file ->
                try {
                    val body = TransportBody.fromJson(file.readText())
                    SignalFile(file, body)
                } catch (e: Exception) {
                    logger.error("Failed to read signal file: ${file.name}", e)
                    file.delete()
                    null
                }
            } ?: emptyList()
    }

    fun deleteFile(signalFile: SignalFile) {
        try {
            signalFile.file.delete()
        } catch (e: Exception) {
            logger.error("Failed to delete signal file: ${signalFile.file.name}", e)
        }
    }

    fun enforceStorageLimits() {
        try {
            // Delete old files
            val now = System.currentTimeMillis()
            val allFiles = getAllFiles()
            for (file in allFiles) {
                if (now - file.lastModified() > config.maxFileAgeMs) {
                    file.delete()
                    logger.debug("Deleted expired signal file: ${file.name}")
                }
            }

            // Enforce total size limit
            var totalSize = getAllFiles().sumOf { it.length() }
            val sortedFiles = getAllFiles().sortedBy { it.lastModified() }
            for (file in sortedFiles) {
                if (totalSize <= config.maxDiskUsageBytes) break
                val fileSize = file.length()
                file.delete()
                totalSize -= fileSize
                logger.debug("Deleted signal file to enforce size limit: ${file.name}")
            }
        } catch (e: Exception) {
            logger.error("Error enforcing storage limits", e)
        }
    }

    private fun getAllFiles(): List<File> {
        val signalFiles = signalsDir.listFiles()?.toList() ?: emptyList()
        val crashFiles = crashDir.listFiles()?.toList() ?: emptyList()
        return signalFiles + crashFiles
    }

    fun clear() {
        signalsDir.listFiles()?.forEach { it.delete() }
        crashDir.listFiles()?.forEach { it.delete() }
    }
}
